docs/vm/table-find.md

```diff
@@ -0,0 +1,182 @@
+Since `getRecord` needs to return a value and the main thread needs to wait for the result, we need to process the stream from `gen.tables`.
+At the same time, `filter` is not currently executing and will not be reading from the data.
+To make this situation work, we would need to buffer the messages from `gen.tables` in memory so that they can be available to be read by `filter` once filter starts executing.
+An alternative is we can duplicate the `gen.tables` node and construct the following plan graphs instead.
+
+```mermaid
+flowchart TB
+    subgraph tableFind
+        a["gen.tables"]-->b["getRecord"]
+    end
+    
+    subgraph yield
+        c["gen.tables"]-->d["filter"]
+        d-->e["yield"]
+    end
+```
+
+Unfortunately, these plan graphs do not accurately represent the code and it leads to an inconsistency in how we handle table streams.
+This is the method currently employed by the query engine, although it is mostly an accident.
+We could decide to make this the explicit behavior if we wanted to and decided that this is the proper implementation.
+
+There is one other problem with these methods.
+It is best illustrated by an IR of the flux code.
+
+```
+define void @main() {
+    %1 = call @gen.tables, n: 100
+    %2 = call @getRecord, tables: %1, idx: 0
+    %3 = call @filter, tables: %1, fn: <funcref>
+    yield %3
+}
+```
+
+You may notice the problem at the call to `getRecord` on line 2 of the IR.
+The value `%2` must be materialized at that time, but we do not know the full plan graph until the end of the function when the variables go out of scope.
+It is not trivial on line 2 to know if `%1` will be live after the call to `getRecord` so we are working with an incomplete plan graph.
+This means we do not know _if_ we have to buffer the messages coming from `gen.tables` until we get to line 3, but we have already been forced to materialize the data on line 2.
+
+Keeping this in mind, we can formulate a few possible solutions.
+
+### Duplicate Nodes in Different Plans
+
+The most trivial solution to this is to call `gen.tables` twice in the two different plan graphs demonstrated above.
+If we duplicate the nodes that span multiple plans, we don't end up with any conflict in the engine itself.
+
+We do create a consistency problem where table streams now have their invocation determined by how they are processed rather than always acting the same way everywhere.
+This is the solution we currently employ and we could continue to employ it but in a more formal capacity.
+
+### Disallow Spanning Plan Graphs
+
+We can decide that it is impossible for nodes to span multiple plan graphs and return an error if someone attempts to do this.
+This would make the query above return an error and would fix the consistency problem.
+It also likely requires the least amount of work and has the most safety because implicit buffering wouldn't be required and we wouldn't need to resolve the liveness problem mentioned above.
```

Comment:
Expanding on this:

While having something automatic would be ideal, I suspect that it may be to complicated for first draft. Adding a simple `buffer()` function for users to explicitly opt-in to plans being reused is simpler (as it relies on users to mark the points of re-use), forwards compatible )there is nothing stopping us from adding automatic buffering later) and has little wasted work (the buffering added should be re-usable for an automatic approach).

## Against the static approach

I think this may end up too pessimistic and will force to buffer too often as well as it may be unclear when it buffers. Doing it for a single function may be possible fairly accurately, but once a stream is passed to another function we then need knowledge of whether that function, or any functions it calls recursively, could duplicate the stream.

```flux
doTwice = (f, x) => {
    f(x)
    f(x)
}

doTwice(f: (x) => ..., x: stream)
```
Still possible to do something in cases like this but I fear it may balloon in complexity with confusion around when and where buffers are added.

## Against the dynamic approach

Another solution we discussed is to use reference counting on the streams. If a stream is evaluated with a reference count of one, we don't need any buffer as no other part of the program requires the stream. If the reference count is higher then we automatically buffer any data read so that the data is not lost.

This solution, while potentially more accurate, is also complicated. It requires accurate reference counting to work which requires accurate analysis to determine when a variable is no longer used (Largely this mirrors Rust's move semantics).

Reply:
In case it resonates, gremlin calls this operation a `barrier` step. https://tinkerpop.apache.org/docs/current/reference/#barrier-step

I'm not sure either `buffer` or `barrier` is the best user-facing name for this, but wanted to offer another term just in case.